src/native/http_connection.c

```diff
@@ -0,0 +1,441 @@
+    assert(s_jni_http_callback_handler.onResponseComplete);
+
+    s_jni_http_callback_handler.sendOutgoingBody = (*env)->GetMethodID(env, cls, "sendOutgoingBody", "(Ljava/nio/ByteBuffer)Z");
+    assert(s_jni_http_callback_handler.sendOutgoingBody);
+}
+
+
+static struct {
+    jfieldID name;
+    jfieldID value;
+} s_http_header_fields;
+
+void s_cache_http_header_handler(JNIEnv *env) {
+    jclass cls = (*env)->FindClass(env, "software/amazon/awssdk/crt/http/HttpRequest$Header");
+    assert(cls);
+    s_http_header_fields.name = (*env)->GetFieldID(env, cls, "name", "Ljava/lang/String;");
+    assert(s_http_header_fields.name);
+
+    s_http_header_fields.value = (*env)->GetFieldID(env, cls, "value", "Ljava/lang/String;");
+    assert(s_http_header_fields.value);
+}
+
+
+
+static void s_on_http_conn_setup(struct aws_http_connection *connection, int error_code, void *user_data) {
+    struct http_jni_connection *http_jni_conn = (struct http_jni_connection *) user_data;
+
+    // Save the native pointer
+    http_jni_conn->native_http_conn = connection;
+
+    // Call the Java Object's "onComplete" callback
+    if (http_jni_conn->java_http_conn) {
+        JNIEnv *env = aws_jni_get_thread_env(http_jni_conn->jvm);
+        (*env)->CallVoidMethod(
+            env, http_jni_conn->java_http_conn, s_http_connection.on_connection_complete, error_code);
+    }
+
+}
+
+static void s_on_http_conn_shutdown(struct aws_http_connection *connection, int error_code, void *user_data) {
+    struct http_jni_connection *http_jni_conn = (struct http_jni_connection *) user_data;
+
+    // Call the Java Object's "onShutdown" callback
+    if (http_jni_conn->java_http_conn) {
+        JNIEnv *env = aws_jni_get_thread_env(http_jni_conn->jvm);
+        (*env)->CallVoidMethod(
+            env, http_jni_conn->java_http_conn, s_http_connection.on_connection_shutdown, error_code);
+    }
+
+}
+
+
+
+/**
+ * Create a new aws_http_request_options struct with default values
+ */
+JNIEXPORT long JNICALL
+    Java_software_amazon_awssdk_crt_http_HttpConnection_httpConnectionNew(JNIEnv *env,
+                                                                          jclass jni_class,
+                                                                          jobject http_conn_jobject,
+                                                                          jlong jni_client_bootstrap,
+                                                                          jlong jni_socket_options,
+                                                                          jlong jni_tls_ctx,
+                                                                          jstring jni_endpoint,
+                                                                          jshort jni_port)
+{
+    struct aws_client_bootstrap *client_bootstrap = (struct aws_client_bootstrap *) jni_client_bootstrap;
```

Comment:
Is there a way to make this safer? You're passing a memory address from java to c, casting it to a struct pointer, and then using it. Or is the norm for JNI code?

Reply:
That's how we've done all our bindings. There's no external reference API and no destructor pattern for external references in JNI. This also allows us to control native memory natively. If we marshalled this into a structure, we'd be subject to GC destruction timing/ordering problems.